oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/osh/OSHEntityTimeUtils.java

```diff
@@ -0,0 +1,357 @@
+    return getModificationTimestamps(osh, true);
+  }
+
+  /**
+   * Returns the list of timestamps at which this entity was modified.
+   *
+   * <p>
+   * If the parameter "recurse" is set to true, it will also include modifications of the object's
+   * child elements (useful to find out when the geometry of this object has been altered).
+   * </p>
+   *
+   * @param osh the osh entity to work on
+   * @param recurse specifies if times of modifications of child entities should also be returned or
+   *        not
+   * @return a list of timestamps where this entity has been modified
+   */
+  public static List<OSHDBTimestamp> getModificationTimestamps(OSHEntity osh, boolean recurse) {
+    switch (osh.getType()) {
+      case NODE:
+        return getModificationTimestamps((OSHNode) osh);
+      case WAY:
+      case RELATION:
+        return getModificationTimestamps(osh, recurse, null);
+      default:
+        throw new UnsupportedOperationException(format(UNSUPPORTED_OSMTYPE_MESSAGE, osh.getType()));
+    }
+  }
+
+  private static List<OSHDBTimestamp> getModificationTimestamps(OSHNode osh) {
+    List<OSHDBTimestamp> result = new ArrayList<>();
+    for (OSMEntity osm : osh.getVersions()) {
+      result.add(osm.getTimestamp());
+    }
+    return Lists.reverse(result);
+  }
+
+  /**
+   * Returns all timestamps at which this entity (or one or more of its child entities) has been
+   * modified and matches a given condition/filter.
+   *
+   * <p>
+   * Consecutive modifications from a single changeset are grouped together (only the last
+   * modification timestamp of the corresponding changeset is considered). This can reduce the
+   * amount of geometry modifications by a lot (e.g. when sequential node uploads of a way
+   * modification causes many intermediate modification states), making results more
+   * "accurate"/comparable as well as allowing faster processing of geometries.
+   * </p>
+   *
+   * @param osh the osh entity to work on
+   * @param osmEntityFilter only timestamps for which the entity matches this filter are returned
+   * @param changesetTimestamps association between timestamps and changeset-ids, can be obtained
+   *        from oshEntity by calling {@link #getChangesetTimestamps}.
+   * @return a list of timestamps where this entity has been modified
+   */
+  public static List<OSHDBTimestamp> getModificationTimestamps(OSHEntity osh,
+      Predicate<OSMEntity> osmEntityFilter, Map<OSHDBTimestamp, Long> changesetTimestamps) {
+    List<OSHDBTimestamp> allModificationTimestamps = getModificationTimestamps(osh, true, osmEntityFilter);
+
+    if (allModificationTimestamps.size() <= 1) {
+      return allModificationTimestamps;
+    }
+    // group modification timestamps by changeset
+    List<OSHDBTimestamp> result = new ArrayList<>(allModificationTimestamps.size());
+    allModificationTimestamps = Lists.reverse(allModificationTimestamps);
```

Comment:
`getModifcationTimestamps(..)` reversed the list of timestamp, and here we reverse it back, maybe we should not reverse in the first place or let's pass parameter in which order we need the modificationTimestamps, so that there is no back and forth.